lib/charms/dns_transfer/v0/dns_transfer.py

```diff
@@ -0,0 +1,532 @@
+which new DNS data has been added or updated.
+
+The DNSTransferRequires provides an `update_relation_data` method to update the relation data by
+passing a list of IP addresses to the provider.
+
+### Provider Charm
+
+Following the previous example, this is an example of the provider charm.
+
+```python
+from charms.dns_transfer.v0.dns_transfer import DNSTransferProvides
+
+class DNSTransferProviderCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.dns_secondary = DNSTransferProvides(self)
+        self.framework.observe(self.dns_secondaries.on.dns_transfer_requirer_data_available, self._handler)
+        ...
+
+    def _handler(self, event: DNSTransferRequirerDataAvailableEvent) -> None:
+        addresses = event.addresses
+        # or self.dns_secondary.get_addresses()
+
+```
+The DNSTransferProvides object wraps the list of relations into a `relations` property
+and provides an `update_relation_data` method to update the relation data by passing
+parameters with data expected by the requirer.
+
+```python
+class DNSTransferProviderCharm(ops.CharmBase):
+    ...
+
+    def _on_config_changed(self, _) -> None:
+        params = {
+            "addresses": ["10.20.30.40", "50.60.70.80"],
+            "transport": "tls",
+            "remote_hostname": "example.com",
+            "zones": ["zone1.com", "zone2.com"],
+        }
+        for relation in self.model.relations[self.dns_secondary.relation_name]:
+            self.dns_secondary.update_relation_data(relation, **params)
+
+```
+"""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "908bcd1f0ad14cabbc9dca25fa0fc87c"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 5
+
+PYDEPS = ["pydantic>=2"]
+
+# pylint: disable=wrong-import-position
+import ipaddress
+import json
+import logging
+from enum import Enum
+from typing import Any, List, Optional, cast
+
+import ops
+import pydantic
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "dns-transfer"
+
+
+class TransportSecurity(str, Enum):
+    """Represent the transport security values.
+
+    Attributes:
+        TCP: tcp
+        TLS: tls
+    """
+
+    TCP = "tcp"
+    TLS = "tls"
+
+
+class DNSTransferProviderData(pydantic.BaseModel):
+    """List of information for the provider to manage and transfer.
+
+    Attributes:
+        addresses: IP list of the units composing the provider's deployment.
+        transport: Type of transport (tls or tcp).
+        remote_hostname: Advertised name of the host in the TLS certificate (optional).
+        zones: List of zone names.
+    """
+
+    addresses: list[pydantic.IPvAnyAddress]
+    transport: TransportSecurity
+    remote_hostname: Optional[str] = None
```

Comment:
You could eventually validate the hostname using [RFC952](https://datatracker.ietf.org/doc/html/rfc952): 
```
<hname> ::= <name>*["."<name>]
<name>  ::= <let>[*[<let-or-digit-or-hyphen>]<let-or-digit>]
```

Reply:
Good point. Added.

Reply:
I'm thinking about renaming it validate_hostname_list on my side since validating a zone would mean validating the records attached to it also. WDYT ?

Reply:
No problem, we can have different names for now and then using the validate_hostname_list in the future validator library.